File: 力扣/学习计划/算法入门/023_全排列/main.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> permute(int[] nums) {
        boolean[] visited = new boolean[nums.length];
        dfs(nums, new ArrayList<>(), visited);
        return res;
    }

    void dfs(int[] nums, List<Integer> path, boolean[] visited) {
        if (path.size() == nums.length) {
            res.add(new ArrayList<>(path));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (visited[i]) continue;
            // 标记已访问
            visited[i] = true;
            path.add(nums[i]);
            // 继续搜索
            dfs(nums, path, visited);
            // 回溯
            visited[i] = false;
            path.remove(path.size() - 1);
        }
    }
}

File: 力扣/学习计划/算法入门/023_全排列/main.ts
/**
 * https://leetcode-cn.com/problems/permutations/
 * 全排列
 */

let res: number[][]
function permute(nums: number[]): number[][] {
    res = new Array();
    let visited = new Array();
    dfs(nums, new Array(), visited)
    return res;
};

const dfs = function (nums: number[], path: number[], visited: boolean[]) {
    if (path.length == nums.length) {
        res.push([...path]);
        return;
    }
    for (let i = 0; i < nums.length; i++) {
        if (visited[i]) continue;
        // 标记已访问
        visited[i] = true;
        path.push(nums[i]);
        // 继续搜索
        dfs(nums, path, visited)
        // 回溯
        visited[i] = false;
        path.pop()
    }
}